clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3269,6 +3279,51 @@ static void handleIntelReqdSubGroupSize(Sema &S, Decl *D,
   S.AddIntelReqdSubGroupSize(D, AL, E);
 }
 
+IntelNamedSubGroupSizeAttr *
+Sema::MergeIntelNamedSubGroupSizeAttr(Decl *D,
+                                      const IntelNamedSubGroupSizeAttr &A) {
+  if (checkAttrMutualExclusion<IntelReqdSubGroupSizeAttr>(*this, D, A))
+    return nullptr;
+  if (checkAttrMutualExclusion<SYCLSimdAttr>(*this, D, A))
+    return nullptr;
+
+  // Check to see if there's a duplicate attribute with different values
+  // already applied to the declaration.
+  if (const auto *DeclAttr = D->getAttr<IntelNamedSubGroupSizeAttr>()) {
+    if (DeclAttr->getType() != A.getType()) {
+      Diag(DeclAttr->getLoc(), diag::warn_duplicate_attribute) << &A;
+      Diag(A.getLoc(), diag::note_previous_attribute);
+    }
+    return nullptr;
+  }
+
+  return IntelNamedSubGroupSizeAttr::Create(Context, A.getType(), A);
+}
+
+static void handleIntelNamedSubGroupSize(Sema &S, Decl *D,
+                                         const ParsedAttr &AL) {
+  if (checkAttrMutualExclusion<IntelReqdSubGroupSizeAttr>(S, D, AL))
+    return;
+  if (checkAttrMutualExclusion<SYCLSimdAttr>(S, D, AL))
+    return;
+
+  if (!AL.isArgIdent(0)) {
```

Comment:
Design question: do you want to also accept something like `[[intel::named_sub_group_size("automatic")]]` where the user passes a string literal? If so, you can add an `else` clause here to call `checkStringLiteralArgumentAttr()`.

Some attributes do this (like `callable_when`) and some don't (like `param_typestate`), so it's up to you whether you care about this or not. I don't have a strong opinion, just wanted to raise the question.

Reply:
I don't think so?  It is nice to know we can relax the spec though, likely in a followup to the current spec.  @Pennycook FYI.

Reply:
Hm.  I don't think this is something we considered.  Thanks for bringing it up.  @AaronBallman, are you aware of any advantages either way?

@rolandschulz and @gmlueck -- any preference here?

Reply:
Assuming we keep both `[[intel::named_sub_group_size()]]` and `[[intel::sub_group_size()]]`, I think we should not accept the string literal, simply to avoid having two ways of doing the same thing.

On the other hand, this might be an opportunity to simplify the extension.  We could get rid of `[[intel::named_sub_group_size()]]` and have only `[[intel::sub_group_size()]]`.  Allowable parameters would be:

* A compile time `constexpr` integral expressions, or
* The string literal `"auto"`, or
* The string literal `"primary"`.

We only added `[[intel::named_sub_group_size()]]` because we needed a way to avoid ambiguity with `[[intel::sub_group_size(primary)]]`.  What if the application happens to have a variable named `primary`?  Is this a reference to that variable, or is it the literal token `primary`.  By requiring it to be enclosed in quotes (`"primary"`), we avoid that ambiguity.

BTW, it looks like the implementation accepts `automatic` not `auto`.  Is that an intentional change?  The spec currently says it should be `auto`.

Reply:
>BTW, it looks like the implementation accepts automatic not auto. Is that an intentional change? The spec currently says it should be auto.

@Pennycook just submitted a patch to change it to `automatic`, since `auto` is a keyword, which confuses our lexer.  We can swap it back to 'auto' if we put it into a string.

I'm ok with re-using the sub_group_size, and think that would be fine (and it could even accept the identifier I think).  If you and @Pennycook  can clarify the spec to be as you want, we can rework this.  

>[[intel::sub_group_size(primary)]]. What if the application happens to have a variable named primary?
Typically we consider this the expression first (that is, the expression 'primary'), but I think it is something we can decide when parsing.

Reply:
> @AaronBallman, are you aware of any advantages either way?

I think the advantages are minimal but the effort to support it is as well. Basically, if we let users use string literals in addition to bare identifiers, it gives users an out who have a macro named `automatic` or `primary`, and it makes users who (mistakenly) think attribute arguments must always be a literal feel more comfortable. The downsides are: it's a few more lines of implementation work and people may wonder why there's two different ways to do the same thing.

> On the other hand, this might be an opportunity to simplify the extension.

I like this way of thinking! A related question I have: do we need to add `[[intel::sub_group_size]]` or could we modify the existing `[[intel::reqd_sub_group_size]]`? That attribute accepts an integer constant expression already, but we could extend it to accept an identifier (and/or string literal)? (I have not spent long thinking about the implementation burden of such an approach, btw.)

Reply:
@erichkeane -- we got our answer via an email discussion. It is okay to support string literals in the attribute (in tandem with using enumeration identifiers) but that it’s not a requirement to support them if it turns out to be an implementation issue. We won't merge `[[intel::named_sub_group_size()]]` and `[[intel::sub_group_size()]]` together into a single attribute.

Reply:
the original was not enum, they were identifiers.  Do they mean enums, or do they mean identifers?

Reply:
Sorry, that may be me losing some nuance in translation. By enum, I mean using an `EnumArgument` to specify the possible values in Attr.td, which allows the user to specify the enumeration constants via an identifier. e.g., `[[intel::named_sub_group_size(automatic)]]`

Reply:
Ah, I see.  Thanks!  So this patch is more or less ok (other than accepting strings if we can).  I'll hopefully get a chance to revisit it later this week.